common/src/View/HandleDragTracker.h

```diff
@@ -0,0 +1,248 @@
+/*
+ Copyright (C) 2021 Kristian Duske
+
+ This file is part of TrenchBroom.
+
+ TrenchBroom is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "FloatType.h"
+#include "Model/HitFilter.h"
+#include "Renderer/Camera.h"
+#include "View/DragTracker.h"
+#include "View/InputState.h"
+
+#include <vecmath/forward.h>
+#include <vecmath/vec.h>
+
+#include <functional>
+#include <iosfwd>
+#include <memory>
+#include <optional>
+#include <type_traits>
+
+namespace TrenchBroom {
+    namespace View {
+        class Grid;
+
+        struct DragState {
+            vm::vec3 initialHandlePosition;
+            vm::vec3 currentHandlePosition;
+            vm::vec3 handleOffset;
+        };
+
+        bool operator==(const DragState& lhs, const DragState& rhs);
+        bool operator!=(const DragState& lhs, const DragState& rhs);
+        std::ostream& operator<<(std::ostream& str, const DragState& dragState);
+
+
+        using GetHandlePosition = std::function<std::optional<vm::vec3>(const InputState&, const DragState&)>;
+
+        enum class ResetInitialHandlePosition {
+            Keep,
+            Reset
+        };
+
+        struct UpdateDragConfig {
+            GetHandlePosition getHandlePosition;
+            ResetInitialHandlePosition resetInitialHandlePosition;
+        };
+
+        /**
+         * The status of a drag. This is returned from a handle drag tracker's delegate when it reacts to a drag event.
+         */
+        enum class DragStatus {
+            /** The drag should continue. */
+            Continue,
+            /** The drag should continue, but the current event could not be applied to the object being dragged. */
+            Deny,
+            /** The drag should be cancelled. */
+            Cancel
+        };
+
+        struct HandleDragTrackerDelegate {
```

Comment:
It might be worth adding a few doc comments to the interface here, because it's what clients need to implement.

Reply:
I will make a final pass with more documentation and tests when all tools have been changed. I keep running into situations where I still need to modify the new tools / delegates to suit some behavior that a controller relies upon.

Reply:
that's fine. I was thinking you were done with the PR - I got a review request on github. Did you send that or was it a github bug?

Reply:
Weird! I didn't want to send one -- but I might have clicked it by accident?

Reply:
I'll ping you again when I'm done with the PR.

Reply:
Haha, this is what I saw so I assumed it was a re-request to review. Maybe it actually just means there was activity on a PR which I was previously requested to review.. idk. I guess it was worthwhile to check up on this anyways ;)

![Capture](https://user-images.githubusercontent.com/239161/120905423-23a8b480-c60f-11eb-89d8-f43e0e91e734.PNG)